app/src/main/kotlin/GridWorld/GridworldMDP.kt

```diff
@@ -29,47 +32,48 @@ class GridworldMDP(
     }
 
     override fun reward(previousState: GridworldState?, action: GridworldAction?, state: GridworldState): Double {
-        var reward = 0.0
         for (r in rewards) {
             if (r.equals(state)) {
-                reward += r.value
+                return r.value
             }
         }
 
-        return reward
+        return 0.0
     }
 
     override fun transition(state: GridworldState, action: GridworldAction): GridworldState {
         if (state.isTerminal) {
             return state
         }
         else if (rewards.any { r -> r.equals(state)}) {
-            return GridworldState(state.x, state.y, true)
+            return _rewardStates.getOrDefault(state, GridworldState(state.x, state.y, true))
```

Comment:
Can you explain the rationale for why .getOrDefault() improves runtime over just returning the GridworldState?

Reply:
This way we don't create a new GridworldState for each transition. Creating a new object has a measurable performance penalty. Returning the states cached via _rewardStates is bettter.